plugin/metrics.go

```diff
@@ -1,43 +1,201 @@
+const (
+	envVarMetricsExporter     = "OTEL_METRICS_EXPORTER"
+	envVarOTLPProtocol        = "OTEL_EXPORTER_OTLP_PROTOCOL"
+	envVarOTLPMetricsProtocol = "OTEL_EXPORTER_OTLP_METRICS_PROTOCOL"
+)
+
+const (
+	otlpProtocolNameGRPC = "grpc"
+	otlpProtocolNameHTTP = "http/protobuf"
+)
+
+const (
+	metricPluginName     = "otel-metrics"
+	metricsPluginVersion = "v0.8.1"
+)
+
+type shutDownMeterProvider interface {
+	metricapi.MeterProvider
+	Shutdown(ctx context.Context) error
+}
+
+var _ shutDownMeterProvider = (*noopShutdownMeterProvider)(nil)
+
+type noopShutdownMeterProvider struct {
+	metricapi.MeterProvider
+}
+
+func (n *noopShutdownMeterProvider) Shutdown(_ context.Context) error {
+	return nil
+}
 
+var _ plugin.Plugin = (*MetricsPlugin)(nil)
+
+// MetricsPlugin configures the OpenTelemetry metrics exporter into the
+// default MeterProvider which can be retrieved via otel.GetMeterProvider.
+//
+// See: https://pkg.go.dev/github.com/ipfs/kubo/plugin#Plugin
 type MetricsPlugin struct {
-	exporter *datadog.Exporter
+	shutDownMeterProvider
+	exporters []metric.Exporter
 }
 
+// Name returns the name of the plugin.
 func (m MetricsPlugin) Name() string {
-	return "datadog-metrics"
+	return metricPluginName
 }
 
+// Version returns the version of the plugin.
 func (m MetricsPlugin) Version() string {
-	return "0.0.1"
+	return metricsPluginVersion
 }
 
-func (m *MetricsPlugin) Init(env *plugin.Environment) error {
-	dd, err := datadog.NewExporter(datadog.Options{
-		Namespace: "kubo",
-	})
+// Init initializes the kubo plugin (from environment variables).
+func (m *MetricsPlugin) Init(_ *plugin.Environment) error {
+	mp, exps, err := newMeterProvider(context.Background())
 	if err != nil {
-		return fmt.Errorf("failed to create the Datadog exporter: %v", err)
+		return err
 	}
 
-	m.exporter = dd
-
-	view.RegisterExporter(dd)
+	m.shutDownMeterProvider = mp
+	m.exporters = exps
+	otel.SetMeterProvider(mp)
+	otel.SetTextMapPropagator(autoprop.NewTextMapPropagator())
 
 	return nil
 }
 
+// Close safely shuts down the plugin.
 func (m *MetricsPlugin) Close() error {
-	m.exporter.Stop()
-	return nil
+	var err error
+
+	// for _, exp := range m.exporters {
+	// 	err = errors.Join(err, exp.ForceFlush(context.Background()))
+	// }
+
+	return errors.Join(err, m.Shutdown(context.Background()))
+}
+
+func newMeterProvider(ctx context.Context) (shutDownMeterProvider, []metric.Exporter, error) {
+	exporters, err := newMetricsExporters(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(exporters) < 1 {
+		return &noopShutdownMeterProvider{noop.NewMeterProvider()}, nil, nil
+	}
+
+	options := []metric.Option{}
+	for _, exporter := range exporters {
+		options = append(options, metric.WithReader(metric.NewPeriodicReader(exporter)))
+	}
+
+	res, err := resource.Merge(
+		resource.Default(),
+		resource.NewSchemaless(
+			semconv.ServiceNameKey.String("kubo"),
+			semconv.ServiceNamespaceKey.String("ipfs"),
+			semconv.ServiceVersionKey.String("v0.20.0"),
```

Comment:
that is going to be fairly annoying to maintain ... can't we just get that version from the tag/annotation we already have? We should have a full `version` in there

Reply:
Version as ldflags?

Reply:
ldflags would work, but in my experience it's quite annoying. Nowadays golang collect info directly from git (commit sha notably), but we'd want the complex version format with both kubo hash and packaging hash anyway.
It's way simpler to tag the workload with the same string as the image tag we run.